driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java

```diff
@@ -285,7 +286,11 @@ static <D, T> void createReadCommandAndExecuteAsync(
 
     static <R> AsyncCallbackSupplier<R> decorateReadWithRetriesAsync(final RetryState retryState, final OperationContext operationContext,
             final AsyncCallbackSupplier<R> asyncReadFunction) {
-        return new RetryingAsyncCallbackSupplier<>(retryState, CommandOperationHelper::chooseRetryableReadException,
+        BinaryOperator<Throwable> onAttemptFailure =
+                (@Nullable Throwable previouslyChosenException, Throwable mostRecentAttemptException) ->
+                        CommandOperationHelper.onRetryableReadAttemptFailure(
+                                operationContext, previouslyChosenException, mostRecentAttemptException);
+        return new RetryingAsyncCallbackSupplier<>(retryState, onAttemptFailure,
```

Comment:
I find it difficult to understand what is happening in these methods, though I am unsure why. I think part of it is the amount of boilerplate: `onAttemptFailure = (priorException, currentException) ->`, and then inlining, might help.

The other, more objective issue is that `onAttemptFailure` is being used as a `failedResultTransformer`, but I would expect such a thing to have no side effects (such as deprioritizing a server in OperationContext).

Reply:
I [updated the docs](https://github.com/mongodb/mongo-java-driver/pull/1367/commits/171d67ed767a65156a768c9667488b13d11049eb) to reflect that side effects are allowed. The key to them being allowed is that the operator is guaranteed to be called once per failed attempt.

> I find it difficult to understand what is happening in these methods...
> ...inlining, might help...

Could you please clarify what you propose to inline and where?

Reply:
I am imagining at least something like:
```
    static <R> AsyncCallbackSupplier<R> decorateReadWithRetriesAsync(final RetryState retryState, 
            final OperationContext operationContext, final AsyncCallbackSupplier<R> asyncReadFunction) {
        return new RetryingAsyncCallbackSupplier<>(retryState,
                onRetryableReadAttemptFailure(operationContext),
                CommandOperationHelper::shouldAttemptToRetryRead,
                logRetryAndGet(retryState, operationContext, asyncReadFunction));
    }
```
This is optional for this PR, and perhaps should be undertaken separately.

Reply:
It looks like a code extraction into a method, rather than inlining. Done in https://github.com/mongodb/mongo-java-driver/pull/1367/commits/69b78ffad40e07040c5f6c8c7e1b9d9e549bdc25.